Add more SignedXml algorithms tests

krwq opened the pull request:
Almost completely fixes https://github.com/dotnet/corefx/issues/16781

the only not tested transformation algorithm is:
urn:mpeg:mpeg21:2003:01-REL-R-NS:licenseTransform";

krwq commented:
@bartonjs are you aware of any x64 only problems? All the x86 builds are passing and all x64 except for windows are failing - I wouldn't expect any x86/x64 differences in SignedXml code but one of the algorithms could be using improper native implementation. HMAC MD5 failures are probably gonna tell us what has happened

krwq commented on src/System.Security.Cryptography.Xml/tests/Samples/EncryptingDecryptingSymmetricKeyWrap.cs:
```diff
@@ -0,0 +1,113 @@
+                var encryptedKey = new EncryptedKey()
+                {
+                    CipherData = new CipherData(EncryptedXml.EncryptKey(innerKey.Key, key)),
+                    EncryptionMethod = new EncryptionMethod(TestHelpers.GetEncryptionMethodName(key, keyWrap: true))
```
this also works when you don't use `keyWrap: true` but it is also not telling about keywrapping in the resulting xml (I don't believe additional spec URL makes sense here - IMO SignedXml should do non-wrapped to wrapped URL conversion here or the URL should be gone)

danmosemsft commented:
@krwq is it rather a non-Windows missing impl problem? Your issue tracks this https://github.com/dotnet/corefx/issues/16585

eg 
```
MESSAGE:
System.PlatformNotSupportedException : Operation is not supported on this platform.
+++++++++++++++++++
STACK TRACE:
at System.Security.Cryptography.MD5CryptoServiceProvider..ctor() in /mnt/resource/j/workspace/dotnet_corefx/master/portablelinux_debug_prtest/bin/obj/Unix.AnyCPU.Debug/System.Security.Cryptography.Csp/netcoreapp/System.Security.Cryptography.Csp.notsupported.cs:line 123
```

Meantime the tests can be disabled against https://github.com/dotnet/corefx/issues/16585

tintoy commented:
LGTM :)

krwq commented:
@danmosemsft for MD5 that it is likely an issue - what about x86/x64 result mismatch?

bartonjs commented:
> what about x86/x64 result mismatch?

There isn't one?  Windows x64 passed.  There's only a Windows vs !Windows difference.

And the test logs show it:

> MESSAGE:
System.PlatformNotSupportedException : Operation is not supported on this platform.
+++++++++++++++++++
STACK TRACE:
at System.Security.Cryptography.RNGCryptoServiceProvider..ctor() 
at System.Security.Cryptography.Xml.SymmetricKeyWrap.TripleDESKeyWrapEncrypt(Byte[] rgbKey, Byte[] rgbWrappedKeyData)

3DES-kw (and probably AES-kw) should switch to RandomNumberGenerator.Create()

krwq commented:
@bartonjs there is only a difference with linux x64 vs x86

bartonjs commented:
@krwq I'm not sure I follow.  We don't have Linux x86.

The table I see is

| job | status |
|----|----|
|Innerloop CentOS7.1 Debug x64 | fail |
|Innerloop CentOS7.1 Release x64 | fail |
|Innerloop OSX10.12 Debug x64  | fail |
|Innerloop OSX10.12 Release x64  | fail |
|Innerloop PortableLinux Debug x64  | fail |
|Innerloop PortableLinux Release x64  | fail |
|Innerloop Ubuntu14.04 Debug x64  | fail |
|Innerloop Ubuntu14.04 Release x64  | fail |
|Innerloop Tizen armel Debug Cross | pass |
|Innerloop Tizen armel Release Cross| pass |
|Innerloop Ubuntu14.04 arm Release | pass |
|Innerloop Ubuntu16.04 arm Debug | pass |
|Innerloop Windows_NT Debug x86 | pass |
|Innerloop Windows_NT Release x64 | pass |
|Vertical netfx Build | pass |
|Vertical uap Build | pass |
|Vertical uapaot Build | pass |
| Windows_NT Debug AllConfigurations Build | pass |

Not sure what the Tizen run runs; but the Linux and macOS tests failed due to RNGCryptoServiceProvider (should be RandomNumberGenerator.Create()) and/or MD5CryptoServiceProvider (should be MD5.Create()).

danmosemsft commented:
@krwq the only Unix that passed is ARM and Tizen and those don't seem to run any tests hence they passed.

krwq commented:
@bartonjs @danmosemsft you're both right I have mis-looked and somehow saw x86 Ubuntu passing and then must've made some general assumption about it. I'll fix it in the morning

krwq commented:
@bartonjs, linux seems to be working, PTAL

bartonjs commented on src/System.Security.Cryptography.Xml/tests/TestHelpers.cs:
```diff
@@ -128,6 +129,61 @@ public static Uri ToUri(string fileName)
+                }
+            }
+
+            throw new ArgumentException("The specified algorithm is not supported for XML Encryption.");
```
Why not say what the algorithm is?

krwq commented on src/System.Security.Cryptography.Xml/tests/Samples/EncryptingDecryptingSymmetricKeyWrap.cs:
```diff
@@ -0,0 +1,113 @@
+
+        public static IEnumerable<object[]> GetSymmetricAlgorithmsPairs()
+        {
+            foreach (var first in TestHelpers.GetSymmetricAlgorithms(skipDes: true))
```
@bartonjs Keywrap does not support combining with DES (there is no keywrap spec string for that either)

bartonjs replied:
Please add a comment in the code to that effect, so that when someone comes across it later they understand it to be a spec reason (this isn't defined) vs a workaround for a test/product reason.

krwq commented on src/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:
```diff
@@ -1301,6 +1302,37 @@ public void VerifyHMAC_SHA384()
+        {
+            // we can't verity the signature if the URI is used
+            SignedXml sign = SignHMAC(moreHmacMD5, new HMACMD5(hmackey), false);
+            Assert.Equal(moreHmacMD5, sign.SignatureMethod);
```
This is matching Desktop

krwq commented:
@bartonjs I believe I addressed all the comments except for naming convention (that would introduce either churn in this PR or inconsistency in this file). PTAL

krwq commented on src/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:
```diff
@@ -1240,18 +1249,21 @@ public void VerifyHMAC_SHA256()
 
             // verify MS-generated signature
-            Assert.True(sign.CheckSignature(new HMACSHA256(hmackey)));
+            Assert.True(sign.CheckSignature(new HMACSHA256(emptyHmacKey)));
```
this is wrong - fix it

krwq replied:
Note: for Verifying signatures I'm still using empty key because it checks it with hardcoded XML and SignXML test does roundtrip checks.

bartonjs replied:
You could also add `Assert.False(sign.CheckSignature(new HMACSHA256(someOtherKey));`

krwq commented on src/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:
```diff
@@ -1268,20 +1280,23 @@ public void VerifyHMAC_SHA512()
 
             // verify MS-generated signature
-            Assert.True(sign.CheckSignature(new HMACSHA512(hmackey)));
+            Assert.True(sign.CheckSignature(new HMACSHA512(emptyHmacKey)));
```
wrong - fix me
